### PR TITLE
fix: porting the install_upgrade.escript from v4.4

### PR DESCRIPTION
**Pull request by terry-xiaoyu**

### CODE REVIEW

File: bin/install_upgrade.escript [hunk 1]
```diff
@@ -359,3 +428,33 @@ erts_vsn() ->
     {ok, Str} = file:read_file(filename:join(["releases", "start_erl.data"])),
     [ErtsVsn, _] = string:tokens(binary_to_list(Str), " "),
     ErtsVsn.
+
+validate_target_version(TargetVersion, TargetNode) ->
+    CurrentVersion = current_release_version(TargetNode),
+    case {get_major_minor_vsn(CurrentVersion), get_major_minor_vsn(TargetVersion)} of
+        {{Major, Minor}, {Major, Minor}} -> ok;
+        _ ->
+            ?INFO("Cannot upgrade/downgrade to ~s from ~s~n"
```
Comment (zmstone) [inline bin/install_upgrade.escript:437]:
```suggestion
            ?INFO("Cannot upgrade/downgrade from ~s to ~s~n"
```

File: bin/install_upgrade.escript [hunk 2]
```diff
@@ -359,3 +428,33 @@ erts_vsn() ->
...
+                  "We only support relup between patch versions",
+                [TargetVersion, CurrentVersion]),
```
Comment (zmstone) [inline bin/install_upgrade.escript:439]:
```suggestion
                [CurrentVersion, TargetVersion]),
```

File: bin/install_upgrade.escript [hunk 3]
```diff
@@ -359,3 +428,33 @@ erts_vsn() ->
...
     [ErtsVsn, _] = string:tokens(binary_to_list(Str), " "),
     ErtsVsn.
+
+validate_target_version(TargetVersion, TargetNode) ->
+    CurrentVersion = current_release_version(TargetNode),
+    case {get_major_minor_vsn(CurrentVersion), get_major_minor_vsn(TargetVersion)} of
+        {{Major, Minor}, {Major, Minor}} -> ok;
+        _ ->
+            ?INFO("Cannot upgrade/downgrade to ~s from ~s~n"
+                  "We only support relup between patch versions",
```
Comment (zmstone) [inline bin/install_upgrade.escript:438]:
```suggestion
                  "Hot upgrade is only supported between patch releases.",
```

File: bin/install_upgrade.escript [hunk 4]
```diff
@@ -287,8 +357,8 @@ permafy(TargetNode, RelName, Vsn) ->
                   make_permanent, [Vsn], ?TIMEOUT),
     ?INFO("Made release permanent: ~p", [Vsn]),
     %% upgrade/downgrade the scripts by replacing them
-    Scripts = [RelNameStr, RelNameStr ++ "_ctl",
-               "nodetool", "install_upgrade.escript"],
+    Scripts = [RelNameStr, RelNameStr++"_ctl", "cuttlefish", "nodetool",
```
Comment (zmstone) [inline bin/install_upgrade.escript:360]:
```suggestion
    Scripts = [RelNameStr, RelNameStr++"_ctl", "nodetool",
```

File: bin/install_upgrade.escript [hunk 5]
```diff
@@ -231,6 +297,10 @@ parse_version(V) when is_list(V) ->
     hd(string:tokens(V,"/")).
 
 check_and_install(TargetNode, Vsn) ->
+    %% Backup the vm.args. VM args should be unchanged during hot upgrade
+    %% but we still backup it here
+    {ok, [[CurrVmArgs]]} = rpc:call(TargetNode, init, get_argument, [vm_args], ?TIMEOUT),
+    {ok, _} = file:copy(CurrVmArgs, filename:join(["releases", Vsn, "vm.args"])),
```
Comment (zmstone) [inline bin/install_upgrade.escript:303]: this file is never used, no need to backup ?

if we do need it after all, we need to add vm_args back to boot script.

File: bin/install_upgrade.escript [hunk 6]
```diff
@@ -209,15 +244,46 @@ find_and_link_release_package(Version, RelName) ->
...
+        {error, Reason} ->
+            ?INFO("create symlink ~p failed", [ReleaseLink]),
+            error({Reason, ReleaseLink})
+    end.
+
+unpack_zipballs(RelNameStr, Version) ->
+    {ok, Cwd} = file:get_cwd(),
+    try
+        GzFile = filename:absname(filename:join(["releases", RelNameStr ++ "-" ++ Version ++ ".tar.gz"])),
+        ZipFiles = filelib:wildcard(filename:join(["releases", RelNameStr ++ "-*" ++ Version ++ "*.zip"])),
```
Comment (zmstone) [inline bin/install_upgrade.escript:269]: 5.0 do not use .zip file for release, tar.gz is enough.

File: bin/install_upgrade.escript [hunk 7]
```diff
@@ -209,15 +244,46 @@ find_and_link_release_package(Version, RelName) ->
...
+        ?INFO("unzip ~p", [ZipFiles]),
+        lists:foreach(
+          fun(Zip) ->
+                  TmdTarD = "/tmp/emqx_untar_" ++ integer_to_list(erlang:system_time()),
+                  ok = filelib:ensure_dir(filename:join([TmdTarD, "dummy"])),
+                  {ok, _} = file:copy(Zip, filename:join([TmdTarD, "emqx.zip"])),
+                  ok = file:set_cwd(filename:join([TmdTarD])),
+                  {ok, _FileList} = zip:unzip("emqx.zip"),
+                  ok = file:set_cwd(filename:join([TmdTarD, "emqx"])),
+                  ok = erl_tar:create(GzFile, filelib:wildcard("*"), [compressed])
```
Comment (zmstone) [inline bin/install_upgrade.escript:279]: this will not work for tar.gz

**Review by zmstone — CHANGES_REQUESTED** [review]
5.0 do not use zip pakcages anymore.
please change (and test) the script to work with .tar.gz packages.

File: build [hunk 8]
```diff
@@ -111,12 +111,12 @@ make_relup() {
...
     local name_pattern
-    name_pattern="${PROFILE}-$(./pkg-vsn.sh "$PROFILE" --vsn_matcher)"
+    name_pattern="${PROFILE}-$(./pkg-vsn.sh "$PROFILE" --vsn_matcher --long)"
     local releases=()
     while read -r tgzfile ; do
         local base_vsn
-        base_vsn="$(echo "$tgzfile" | grep -oE "[0-9]+\.[0-9]+\.[0-9]+(-(alpha|beta)\.[0-9])?(-[0-9a-f]{8})?" | head -1)"
-        tar -C "$rel_dir" -zxf ---keep-old-files "$tgzfile" emqx/releases emqx/lib
+        base_vsn="$(echo "$tgzfile" | grep -oE "[0-9]+\.[0-9]+\.[0-9]+(-(alpha|beta|rc)\.[0-9])?(-[0-9a-f]{8})?" | head -1)"
+        tar -C "$rel_dir" -zxf "$tgzfile" --keep-old-files releases lib
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply build]: I changed the implementation to creating tmp dirs, as the --skip-old-files option of `tar` command is not supported on Mac OSX.

**Review by zmstone — APPROVED** [review]

**Pull request merged**